asserts/internal/grouping.go

```diff
@@ -98,32 +107,50 @@ func (gr *Groupings) AddTo(g *Grouping, group uint16) error {
 		g.elems = []uint16{group}
 		return nil
 	}
-	// TODO: support using a bit-set representation after the size point
-	// where the space cost is the same
-	found, j := g.search(group)
+	found, j := gr.search(g, group)
 	if found {
 		return nil
 	}
+	newsize := g.size + 1
+	if newsize > gr.bitsetThreshold {
```

Comment:
Wonder if it makes sense to just go with a bitset implementation and drop the array one. Looking at #8569 and the code that landed, with 512 groups, we would end up using 32*2 bytes + slice bookkeeping for each asserts refresh. There is quite a loot of other bookkeeping around asserts.Pool that will be close to doubling that. Still, even within ~100 bytes per refresh, the amount of memory feels small compared to other operations that we do.

Reply:
the main reason for not going always with the bit set one is that as the comment on Groupings explain we expect most groupings to be small. space saving is not the main issue, the problem with small groupings and the bit set representation is that iteration is not that great.

Reply:
On the same microoptimizatio note I tried using bits.TrailingZeros/LeadingZeros to trim the inner loop range. Got this:

```
diff --git a/asserts/internal/grouping.go b/asserts/internal/grouping.go
index 924d443456..7136250d13 100644
--- a/asserts/internal/grouping.go
+++ b/asserts/internal/grouping.go
@@ -25,6 +25,7 @@ import (
        "encoding/binary"
        "errors"
        "fmt"
+       "math/bits"
        "sort"
 )
 
@@ -257,19 +258,21 @@ func (gr *Groupings) bitsetIter(g *Grouping, f func(group uint16) error) error {
                if w == 0 {
                        continue
                }
-               bit := uint16(1)
-               for j := uint16(0); j < 16; j++ {
-                       if w&bit != 0 {
-                               if err := f(i*16 + j); err != nil {
-                                       return err
-                               }
-                               c--
-                               if c == 0 {
-                                       // found all elements
-                                       return nil
-                               }
+               firstBit := bits.TrailingZeros16(w)
+               lastBit := 15 - bits.LeadingZeros16(w)
+               for j := uint16(firstBit); j < uint16(lastBit+1); j++ {
+                       bit := uint16(1) << j
+                       if w&bit == 0 {
+                               continue
+                       }
+                       if err := f(i*16 + j); err != nil {
+                               return err
+                       }
+                       c--
+                       if c == 0 {
+                               // found all elements
+                               return nil
                        }
-                       bit <<= 1
                }
        }
        return nil
diff --git a/asserts/internal/grouping_test.go b/asserts/internal/grouping_test.go
index 5a2d5e4490..7f702b059d 100644
--- a/asserts/internal/grouping_test.go
+++ b/asserts/internal/grouping_test.go
@@ -366,6 +366,38 @@ func (s *groupingsSuite) TestBitsetIter(c *C) {
        c.Check(elems, HasLen, 32)
 }
 
+func (s *groupingsSuite) TestBitsetIterSimple(c *C) {
+       gr, err := internal.NewGroupings(32)
+       c.Assert(err, IsNil)
+
+       var elems []uint16
+       f := func(group uint16) error {
+               elems = append(elems, group)
+               return nil
+       }
+
+       var g internal.Grouping
+       err = gr.AddTo(&g, 1)
+       c.Assert(err, IsNil)
+       err = gr.AddTo(&g, 5)
+       c.Assert(err, IsNil)
+       err = gr.AddTo(&g, 17)
+       c.Assert(err, IsNil)
+       err = gr.AddTo(&g, 24)
+       c.Assert(err, IsNil)
+
+       l := gr.Serialize(&g)
+       c.Check(l, DeepEquals,
+               internal.Serialize([]uint16{4,
+                       uint16(1<<1 | 1<<5),
+                       uint16(1<<(17-16) | 1<<(24-16)),
+               }))
+
+       err = gr.Iter(&g, f)
+       c.Assert(err, IsNil)
+       c.Check(elems, DeepEquals, []uint16{1, 5, 17, 24})
+}
+
 func (s *groupingsSuite) TestBitsetIterError(c *C) {
        gr, err := internal.NewGroupings(16)
        c.Assert(err, IsNil)
```

Reply:
Thanks, I'll try to expand the comment about the why the two implementations beyond the size consideration. I'll explore what we discussed about changing the code to shift w to 0, and then adding TrailingZero16.